jax/_src/api_util.py

```diff
@@ -59,7 +58,18 @@ def _ensure_str_tuple(x: Union[str, Iterable[str]]) -> Tuple[str, ...]:
   else:
     return tuple(map(_ensure_str, x))
 
-@lu.transformation_with_aux
+# TODO(frostig,mattjj): consider "upstreaming" this to `linear_util`,
+# so that any transformation wrapped by `transformation_with_aux` can
+# choose to set the `pure` option (for an overwriting store), not only
+# the following flattening transformations. We have no such known uses
+# for now, though.
+def transformation_with_aux(xform):
+  def wrap_xform(*args, pure=False):
+    return lu.transformation_with_aux(xform, pure=pure)(*args)
```

Comment:
Can we change the 'pure' name? I'm not sure it's really descriptive, since these functions are always pure in the functional sense. How about "rerunnable" or "rerunnable_with_equal_aux" or something?

Actually might also be slightly cleaner not to change linear_util.py _or_ api_util.py in this PR, and just define these in custom_transpose.py:

```python
class StoreEqual:
  def store(self, val):
    if self._val is not lu_EMPTY_STORE_VALUE and val != self._val:
      raise lu.StoreException(f"Store assignment mismatch, from {self._val} to {val}")
    self._val = val

  @property
  def val(self):
    ...

def transformation_with_aux(gen, fun: lu.WrappedFun, *gen_static_args) -> Tuple[lu.WrappedFun, Any]:
  out_store = StoreEqual()
  out_thunk = lambda: out_store.val
  return fun.wrap(gen, gen_static_args, out_store), out_thunk
```

That way we don't need to decide on the name of a new kwarg, and we don't have to touch linear_util.py / api_util.py while this stuff is in flux.

WDYT? Anything I'm missing on why we'd want to change linear_util.py / api_util.py here?

(Also, since I've been slow to review, we can just land this and iterate if you prefer.)

Reply:
Great idea. Fixed, namely forked the three necessary functions over in custom_transpose.py.